packages/dicomImageLoader/src/imageLoader/internal/rangeRequest.ts

```diff
@@ -0,0 +1,241 @@
+  const promise = new Promise<{
+    contentType: string;
+    imageFrame: Uint8Array;
+    complete: boolean;
+    loadNextRange: () => any;
+  }>(async (resolve, reject) => {
+    const headers = Object.assign(
+      {},
+      defaultHeaders
+      /* beforeSendHeaders */
+    );
+
+    Object.keys(headers).forEach(function (key) {
+      if (headers[key] === null) {
+        headers[key] = undefined;
+      }
+      if (key === 'Accept' && url.indexOf('accept=') !== -1) {
+        headers[key] = undefined;
+      }
+    });
+
+    try {
+      cornerstone.triggerEvent(
+        cornerstone.eventTarget,
+        'cornerstoneimageloadstart',
+        {
+          url,
+          imageId,
+        }
+      );
+
+      streamCache[imageId] = {
+        initialBytes: initialBytes as number,
+        totalRanges: totalRanges as number,
+        rangesFetched: 0,
+      };
+
+      const { bytes, responseHeaders } = await fetchRangeAndAppend(
+        url,
+        imageId,
+        headers,
+        [0, initialBytes as number]
+      );
+
+      // Resolve promise with the first range, so it can be passed through to
+      // cornerstone via the usual image loading pathway. All subsequent
+      // ranges will be passed and decoded via events.
+      const done = loadIsComplete(imageId);
+      const contentType = responseHeaders.get('content-type');
+      const extract = extractMultipart(contentType, bytes, {}, true);
+
+      resolve({
+        ...extract,
+        complete: done && !retrieveOptions.isLossy,
+        done,
+        isLossy: !!retrieveOptions.isLossy,
+        loadNextRange: done
+          ? undefined
+          : async () => {
+              const loadedBytes = streamCache[imageId].byteArray.length;
+              const totalBytes = streamCache[imageId].totalBytes;
+              if (loadIsComplete(imageId)) {
+                return {
+                  complete: true,
+                  imageFrame: streamCache[imageId].byteArray,
+                };
+              }
+
+              const rangesFetched = streamCache[imageId].rangesFetched;
+              const rangeEnd =
+                Math.ceil(
+                  (totalBytes - loadedBytes) /
+                    ((totalRanges as number) - rangesFetched)
+                ) + loadedBytes;
+
+              const { bytes } = await fetchRangeAndAppend(
+                url,
+                imageId,
+                headers,
+                [loadedBytes, rangeEnd]
+              );
+
+              return {
+                complete: loadIsComplete(imageId),
+                imageFrame: bytes,
+                contentType,
+              };
+            },
+      });
+    } catch (err: any) {
+      errorInterceptor(err);
+      console.error(err);
+      reject(err);
+    }
+  });
+
+  return promise;
+}
+
+function loadIsComplete(imageId) {
+  return (
+    streamCache[imageId].byteArray.length === streamCache[imageId].totalBytes
+  );
+}
```

Comment:
I prefer starting the function name with is/has so `isLoadFinished/completed` or `isStreamComplete` etc